File: src/test/java/no/nav/veilarbaktivitet/aktivitetskort/AktivitetsCompareUtilTest.java
package no.nav.veilarbaktivitet.aktivitetskort;

import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetStatus;
import no.nav.veilarbaktivitet.testutils.AktivitetDataTestBuilder;
import org.assertj.core.api.Assertions;
import org.junit.Test;

import java.util.List;

public class AktivitetsCompareUtilTest {

    @Test
    public void deltakelsestatusendring_skal_være_faktisk_endring() {
        var gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet();
        var nyAktivitet = gammelAktivitet
                .withEksternAktivitetData(gammelAktivitet
                    .getEksternAktivitetData()
                        .withEtiketter(List.of(new Etikett("DELTAR")))
                );
        Assertions.assertThat(
            AktivitetskortCompareUtil
                .erFaktiskOppdatert(gammelAktivitet, nyAktivitet)
        ).isTrue();
    }

    @Test
    public void ider_er_ikke_faktisk_endring() {
        var gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet();
        var nyAktivitet = gammelAktivitet
                .withVersjon(10l)
                .withId(12l);
        Assertions.assertThat(
            AktivitetskortCompareUtil
            .erFaktiskOppdatert(gammelAktivitet, nyAktivitet)
        ).isFalse();
    }

    @Test
    public void forhåndsorientering_er_ikke_faktisk_endring() {
        var gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet();
        var nyAktivitet = gammelAktivitet
                .withForhaandsorientering(AktivitetDataTestBuilder.nyForhaandorientering());
        Assertions.assertThat(
                AktivitetskortCompareUtil
                        .erFaktiskOppdatert(gammelAktivitet, nyAktivitet)
        ).isFalse();
    }

    @Test
    public void aktivitetsstatus_endring_er_ikke_faktisk_endring() {
        var gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
                .withStatus(AktivitetStatus.GJENNOMFORES);
        var nyAktivitet = gammelAktivitet
                .withStatus(AktivitetStatus.FULLFORT);
        Assertions.assertThat(
                AktivitetskortCompareUtil
                        .erFaktiskOppdatert(gammelAktivitet, nyAktivitet)
        ).isFalse();
    }

    @Test
    public void tiltaksnavn_endring_er_faktisk_endring() {
        var gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet();
        var nyAktivitet = gammelAktivitet
            .withEksternAktivitetData(gammelAktivitet.getEksternAktivitetData()
                    .withDetaljer(List.of(new Attributt("tiltaksnavn", "Hurra AS")))
            );
        Assertions.assertThat(
            AktivitetskortCompareUtil
                .erFaktiskOppdatert(gammelAktivitet, nyAktivitet)
        ).isTrue();
    }

}

File: src/test/kotlin/no/nav/veilarbaktivitet/aktivitetskort/AktivitetsCompareUtilTest.kt
package no.nav.veilarbaktivitet.aktivitetskort

import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetStatus
import no.nav.veilarbaktivitet.aktivitet.domain.aktiviteter.AktivitetMuterbareFelter
import no.nav.veilarbaktivitet.aktivitet.domain.aktiviteter.Eksternaktivitet
import no.nav.veilarbaktivitet.aktivitet.domain.aktiviteter.SporingsData
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortCompareUtil.erFaktiskOppdatert
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Attributt
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Etikett
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Sentiment
import no.nav.veilarbaktivitet.person.Innsender
import no.nav.veilarbaktivitet.testutils.AktivitetDataTestBuilder
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.ZonedDateTime

internal class AktivitetsCompareUtilTest {

    private fun AktivitetData.toEndre(): Eksternaktivitet.Endre {
        return Eksternaktivitet.Endre(
            id = id,
            versjon = versjon,
            muterbareFelter = AktivitetMuterbareFelter(
                tittel = tittel,
                beskrivelse = beskrivelse,
                fraDato = fraDato,
                tilDato = tilDato,
                lenke = lenke,
            ),
            sporing = SporingsData(
                endretAv = endretAv,
                endretAvType = endretAvType,
                endretDato = ZonedDateTime.now(),
            ),
            eksternAktivitetData = eksternAktivitetData,
            erAvtalt = isAvtalt,
            status = status,
        )
    }

    @Test
    fun deltakelsestatusendring_skal_være_faktisk_endring() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
        val nyAktivitet = gammelAktivitet.toEndre().copy(
            eksternAktivitetData = gammelAktivitet.eksternAktivitetData.copy(
                etiketter = listOf(
                    Etikett("Deltar", Sentiment.POSITIVE, "DELTAR")
                )
            )
        )
        assertThat(erFaktiskOppdatert(nyAktivitet, gammelAktivitet)).isTrue()
    }

    @Test
    fun ider_er_ikke_faktisk_endring() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
        val nyAktivitet = gammelAktivitet.toEndre().copy(
            versjon = 10L,
            id = 12L,
        )
        assertThat(erFaktiskOppdatert(nyAktivitet, gammelAktivitet)).isFalse()
    }

    @Test
    fun `forhåndsorientering er ikke faktisk endring`() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
            .withForhaandsorientering(AktivitetDataTestBuilder.nyForhaandorientering())
        val nyAktivitet = gammelAktivitet.toEndre()
        assertThat(erFaktiskOppdatert(nyAktivitet, gammelAktivitet)).isFalse()
    }

    @Test
    fun `aktivitetsstatus endring er ikke faktisk endring fordi dette skjer i egne endepunkter og transaksjoner`() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
            .withStatus(AktivitetStatus.GJENNOMFORES)
        val nyAktivitet = gammelAktivitet.toEndre().copy(
            status = AktivitetStatus.FULLFORT,
        )
        assertThat(erFaktiskOppdatert(nyAktivitet, gammelAktivitet)).isFalse()
    }

    @Test
    fun endretTidspunkt_og_endretAv_er_ikke_faktisk_endring() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
            .withEndretAvType(Innsender.NAV)
        val base = gammelAktivitet.toEndre()
        val nyEndretAaAktivitet = base.copy(sporing = base.sporing.copy(endretAv = "Hei"))
        val nyEndretDatoAktivitet = base.copy(sporing = base.sporing.copy(endretDato = ZonedDateTime.now().minusDays(10)))
        val nyEndretTypeAktivitet = base.copy(sporing = base.sporing.copy(endretAvType = Innsender.ARBEIDSGIVER))
        assertThat(erFaktiskOppdatert(nyEndretAaAktivitet, gammelAktivitet)).isFalse()
        assertThat(erFaktiskOppdatert(nyEndretDatoAktivitet, gammelAktivitet)).isFalse()
        assertThat(erFaktiskOppdatert(nyEndretTypeAktivitet, gammelAktivitet)).isFalse()
    }

    @Test
    fun tiltaksnavn_endring_er_faktisk_endring() {
        val gammelAktivitet = AktivitetDataTestBuilder.nyEksternAktivitet()
        val nyAktivitet = gammelAktivitet.toEndre().copy(
            eksternAktivitetData = gammelAktivitet.eksternAktivitetData.copy(
                detaljer = listOf(
                    Attributt("tiltaksnavn", "Hurra AS")
                )
            )
        )
        assertThat(erFaktiskOppdatert(nyAktivitet, gammelAktivitet)).isTrue()
    }
}
